x/oracle/keeper/native_token.go

```diff
@@ -0,0 +1,291 @@
+	newBalance.Block = uint64(ctx.BlockHeight())
+	if amount.IsPositive() {
+		newBalance.Change = types.BalanceInfo_ACTION_DEPOSIT
+	} else {
+		// TODO: check if this validator has withdraw all its asset and then we can move it out from the staker's validatorList
+		// currently when withdraw happened we assume this validator has left the staker's validatorList (deposit/withdraw all of that validator's staking ETH(<=32))
+		newBalance.Change = types.BalanceInfo_ACTION_WITHDRAW
+		for i, vPubkey := range stakerInfo.ValidatorPubkeyList {
+			if vPubkey == validatorPubkey {
+				stakerInfo.ValidatorPubkeyList = append(stakerInfo.ValidatorPubkeyList[:i], stakerInfo.ValidatorPubkeyList[i+1:]...)
+				break
+			}
+		}
+	}
+
+	newBalance.Balance += amount.Int64()
+
+	keyStakerList := types.NativeTokenStakerListKey(assetID)
+	valueStakerList := store.Get(keyStakerList)
+	stakerList.StakerAddrs = make([]string, 0, 1)
+	if valueStakerList != nil {
+		k.cdc.MustUnmarshal(valueStakerList, &stakerList)
+	}
+	exists := false
+	for idx, stakerExists := range stakerList.StakerAddrs {
+		// this should noly happen when do withdraw
+		if stakerExists == stakerAddr {
+			if newBalance.Balance <= 0 {
+				stakerList.StakerAddrs = append(stakerList.StakerAddrs[:idx], stakerList.StakerAddrs[idx+1:]...)
+				valueStakerList = k.cdc.MustMarshal(&stakerList)
+				store.Set(keyStakerList, valueStakerList)
+			}
+			exists = true
+			stakerInfo.StakerIndex = int64(idx)
+			break
+		}
+	}
+	if !exists {
+		if !amount.IsPositive() {
+			return errors.New("remove unexist validator")
+		}
+		stakerList.StakerAddrs = append(stakerList.StakerAddrs, stakerAddr)
+		valueStakerList = k.cdc.MustMarshal(&stakerList)
+		store.Set(keyStakerList, valueStakerList)
+		stakerInfo.StakerIndex = int64(len(stakerList.StakerAddrs) - 1)
+	}
+
+	if newBalance.Balance <= 0 {
+		store.Delete(key)
+	} else {
+		stakerInfo.BalanceList = append(stakerInfo.BalanceList, &newBalance)
+		bz := k.cdc.MustMarshal(stakerInfo)
+		store.Set(key, bz)
+	}
+	eventValue := fmt.Sprintf("%d_%s", stakerInfo.StakerIndex, validatorPubkey)
+	if newBalance.Change == types.BalanceInfo_ACTION_DEPOSIT {
+		eventValue = fmt.Sprintf("%s_%s", types.AttributeValueNativeTokenDeposit, eventValue)
+	} else {
+		eventValue = fmt.Sprintf("%s_%s", types.AttributeValueNativeTokenWithdraw, eventValue)
+	}
+	// emit an event to tell a new valdiator added/or a validator is removed for the staker
+	ctx.EventManager().EmitEvent(sdk.NewEvent(
+		types.EventTypeCreatePrice,
+		sdk.NewAttribute(types.AttributeKeyNativeTokenChange, eventValue),
+	))
+	return nil
+}
+
+// GetStakerInfos returns all stakers information
+func (k Keeper) GetStakerInfos(ctx sdk.Context, assetID string) (ret []*types.StakerInfo) {
+	store := ctx.KVStore(k.storeKey)
+	iterator := sdk.KVStorePrefixIterator(store, types.NativeTokenStakerKeyPrefix(assetID))
+	defer iterator.Close()
+	for ; iterator.Valid(); iterator.Next() {
+		sInfo := types.StakerInfo{}
```

Comment:
Reuse objects StakerInfo  using sync.Pool ?

Reply:
sInfo will be appended to slice as pointer, So the memory should be allocated anyway(need to do deep copy if we use sync.pool).